builder/swig/python.bzl

```diff
@@ -0,0 +1,172 @@
+        local_defines = lib_compilation_context.local_defines,
+        quote_includes = lib_compilation_context.quote_includes,
+        system_includes = depset(
+            [file.dirname for file in ctx.attr._python_header[CcInfo].compilation_context.headers.to_list()],
+            transitive = [lib_compilation_context.system_includes],
+        )
+    )
+
+    return [
+        DefaultInfo(files = depset([wrap_src, wrap_py])),
+        CcInfo(
+            compilation_context = compilation_context,
+            linking_context = ctx.attr.lib[CcInfo].linking_context,
+        ),
+    ]
+
+
+_swig_python_wrapper = rule(
+    implementation = _swig_python_wrapper_impl,
+    attrs = {
+        "lib": attr.label(
+            doc = "The C library for which to generate the wrapper",
+            providers = [CcInfo],
+            mandatory = True,
+        ),
+        "class_name": attr.string(
+            doc = "Optional override for the python class name (default: same as target name)",
+        ),
+        "shared_lib_name": attr.string(
+            doc = "Optional override for the dynamic library name (default: '_' + target name)",
+        ),
+        "interface": attr.label(
+            doc = "Optional SWIG interface (.i) file",
+            allow_single_file = True,
+        ),
+        "includes": attr.label_list(
+            doc = "Additional SWIG files required for wrapper generation",
+            allow_files = True,
+        ),
+        "enable_cxx": attr.bool(
+            doc = "Enable SWIG C++ processing (default: False)",
+            default = True,
+        ),
+        "extra_args": attr.string_list(
+            doc = "Extra arguments to be passed to SWIG",
+        ),
+        "_python_header": attr.label(
+            default = Label("@python39//:python_headers"),
+        ),
+        "_swig": attr.label(
+            default = Label("@swig//:swig"),
+            allow_single_file = True,
+            executable = True,
+            cfg = "host",
+        ),
+    },
+)
+
+
+def swig_python_wrapper(**kwargs):
+    # workaround for select() not being allowed as a default argument
+    # cf. https://github.com/bazelbuild/bazel/issues/287
+    _swig_python_wrapper(
+        **kwargs,
+    )
+
+
+def swig_python(name, lib, shared_lib_name=None, **kwargs):
+    swig_wrapper_name = name + "__swig"
+    if not shared_lib_name:
+        shared_lib_name = "_" + name
+
+    swig_python_wrapper(
+        class_name = name,
+        name = swig_wrapper_name,
+        shared_lib_name = shared_lib_name,
+        lib = lib,
+        **kwargs,
+    )
+
+    def swig_cc_binary(shared_lib_filename):
+        # name doesn't accept select() either
+        native.cc_binary(
+            name = shared_lib_filename,
+            deps = [lib, swig_wrapper_name],
+            srcs = [swig_wrapper_name],
+            linkshared = True,
+        )
+
+    swig_cc_binary_so = swig_cc_binary(shared_lib_name + ".so")
```

Comment:
By having it outside of a `select`, will we always have a `.so` rule instantiated? Even if we only assign it to a variable? I'm not sure what happens in this case. If we always end up with a `.so` rule declared then that's not correct.

Reply:
If we have this declaration inside `select` (twice), we get an error `cc_binary rule 'native_client_python.so' in package 'rust' conflicts with existing cc_binary rule`. It means that all rules in `select` are created (we have checked it). So the `select` block can be replaced with a simple rules creation.